src/metabase/related.clj

```diff
@@ -0,0 +1,274 @@
+
+(defn similarity
+  "How similar are entities `a` and `b` based on a structural comparison of their
+   definition (MBQL).
+   For the purposes of finding related entites we are only interested in
+   context-bearing subforms (field, segment, and metric references). We also
+   don't care about generalizations (less context-bearing forms) and refinements
+   (more context-bearing forms), so we just check if the less specifc form is a
+   subset of the more specific one."
+  [a b]
+  (let [context-a (-> a definition collect-context-bearing-forms)
+        context-b (-> b definition collect-context-bearing-forms)]
+    (/ (count (set/intersection context-a context-b))
+       (max (min (count context-a) (count context-b)) 1))))
+
+(defn- rank-by-similarity
+  [reference entities]
+  (->> entities
+       (remove #{reference})
+       (map #(assoc % :similarity (similarity reference %)))
+       (sort-by :similarity >)))
+
+(defn- interesting-mix
+  "Create an interesting mix of matches. The idea is to have a balanced mix
+   between close (best) matches and more diverse matches to cover a wider field
+   of intents."
+  [matches]
+  (let [[best rest] (split-at max-best-matches matches)]
+    (concat best (->> rest shuffle (take max-serendipity-matches)))))
+
+(defn- metrics-for-table
+  [table]
+  (filter mi/can-read? (db/select Metric
+                         :table_id  (:id table)
+                         :is_active true)))
+
+(defn- segments-for-table
+  [table]
+  (filter mi/can-read? (db/select Segment
+                         :table_id  (:id table)
+                         :is_active true)))
+
+(defn- linking-to
+  [table]
+  (->> (db/select-field :fk_target_field_id Field
+         :table_id (:id table)
+         :fk_target_field_id [:not= nil])
+       (map (comp Table :table_id Field))
+       distinct
+       (filter (every-pred (comp nil? :visibility_type) mi/can-read?))
+       (take max-matches)))
+
+(defn- linked-from
+  [table]
+  (let [fields (db/select-field :id Field :table_id (:id table))]
+    (->> (db/select-field :table_id Field
+           :fk_target_field_id [:in fields])
+         (map Table)
+         (filter (every-pred (comp nil? :visibility_type) mi/can-read?))
+         (take max-matches))))
+
+(defn- cards-sharing-dashboard
+  [card]
+  (when-let [dashboards (not-empty (db/select-field :dashboard_id DashboardCard
+                                     :card_id (:id card)))]
+    (->> (db/select-field :card_id DashboardCard
+           :dashboard_id [:in dashboards]
+           :card_id [:not= (:id card)])
+         (map Card)
+         (filter (every-pred (comp false? :archived) mi/can-read?))
+         (take max-matches))))
+
+(defn- similar-questions
+  [card]
+  (->> (db/select Card
+         :table_id (:table_id card)
+         :archived false)
+       (filter mi/can-read?)
+       (rank-by-similarity card)
+       (filter (comp pos? :similarity))))
+
+(defn- canonical-metric
```

Comment:
why is this a single metric?

Reply:
This tries to catch where you are using an aggregation that also exists as a metric. For instance say you have a metric TotalIncome = [:sum [:field-id 8]]. If you then make a card using the same aggregation, it would be good to suggest you use TotalIncome metric instead.
I haven't fully explored this design space, but I think it could be helpful in fending off entropy.